### PR TITLE
Sort projects by update/add comparator

### PR DESCRIPTION
**Pull request by wow-such-code**
Adds comparator, tests and sorting of projects (their summaries) by the timepoint of last change

### CODE REVIEW

**Comment by wow-such-code** [comment]
[Implement a Comparator for ProjectSummaries that orders them by their latest modification date](https://trello.com/c/voAyKjJN/1013-implement-a-comparator-for-projectsummaries-that-orders-them-by-their-latest-modification-date)

File: sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/components/projectoverview/LastChangedComparator.groovy [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package life.qbic.portal.sampletracking.components.projectoverview
+
+import java.time.Instant
+import java.util.Comparator
+
+/**
+ * Compares ProjectSummaries by the last change time as denoted by the timestamp
```
Comment (jenniferboedker) [inline sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/components/projectoverview/LastChangedComparator.groovy:7]:
```suggestion
 * Compares {@link ProjectSummary} by the last change time as denoted by the timestamp
```

File: sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/components/projectoverview/ProjectSummary.groovy [hunk 2]
```diff
@@ -3,6 +3,7 @@ package life.qbic.portal.sampletracking.components.projectoverview
 import groovy.transform.EqualsAndHashCode
 import java.time.Instant
 import life.qbic.business.project.Project
+import java.util.Comparator
```
Comment (jenniferboedker) [inline sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/components/projectoverview/ProjectSummary.groovy:6]: why do you need it here?

File: sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/components/projectoverview/ProjectOverviewViewModel.groovy [hunk 3]
```diff
@@ -38,6 +38,8 @@ class ProjectOverviewViewModel {
         projectResourceService.iterator().each { Project project ->
             addProject(project)
         }
+        Collections.sort(projectOverviews, new LastChangedComparator(SortOrder.DESCENDING))
+
```
Comment (jenniferboedker) [inline sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/components/projectoverview/ProjectOverviewViewModel.groovy:42]: This seems not to work. I think you need to import the SortOrder. I tested it locally and I got an error

**Review by jenniferboedker — CHANGES_REQUESTED** [review]
Thanks for your contribution. Unfortunately, it does not run. I got a failure upon testing.,

File: sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/components/projectoverview/ProjectSummary.groovy (hunk 2 above)
Comment (wow-such-code) [reply sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/components/projectoverview/ProjectSummary.groovy:6]: we don't, leftover from having the comparator in this file

File: sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/components/projectoverview/ProjectOverviewViewModel.groovy (hunk 3 above)
Comment (wow-such-code) [reply sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/components/projectoverview/ProjectOverviewViewModel.groovy]: should be fixed now

File: sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/samples/SamplesDbConnector.groovy [hunk 4]
```diff
@@ -107,23 +106,25 @@ class SamplesDbConnector implements CountSamplesDataSource, DownloadSamplesDataS
       Connection connection = connectionProvider.connect()
       String latestChangeQuery = "select MAX(UNIX_TIMESTAMP(arrival_time)) from samples_locations where sample_id LIKE ?"
       String sqlRegex = "${projectCode}%"
-      Timestamp latest = new Timestamp(0)
+      Instant latest = Instant.MIN
```
Comment (jenniferboedker) [inline sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/samples/SamplesDbConnector.groovy:109]: What does Instant.Min do? create a instant of a minute to initialize the value?

**Review by jenniferboedker — APPROVED** [review]
Nice, now it works!

File: sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/samples/SamplesDbConnector.groovy (hunk 4 above)
Comment (wow-such-code) [reply sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/samples/SamplesDbConnector.groovy:109]: It's like Integer.MIN_VALUE and guarantees that projects without a recorded change show up last when sorted by that time.

Comment (jenniferboedker) [reply sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/samples/SamplesDbConnector.groovy:109]: very neat, thanks for that explanation!

**Pull request merged**